### PR TITLE
Add Disable instance api for Linux consumption (https://github.com/Azure/azure-functions-host/issues/5032)

### PR DESCRIPTION
**Pull request by balag0**
This will be used to put the container offline before it is killed to reduce the time containers marked for deletion are processing events.

### CODE REVIEW

File: src/WebJobs.Script/Extensions/FileUtility.cs [hunk 1]
```diff
@@ -282,5 +282,19 @@ private static void DoSafeAction(Action action, bool ignoreErrors)
             {
             }
         }
+
+        public static async Task MarkContainerDisabled()
```
Comment (fabiocav) [inline src/WebJobs.Script/Extensions/FileUtility.cs]: These new methods don’t really belong in the FileUtility class. Can we move this to a more appropriate type?

File: src/WebJobs.Script/Utility.cs [hunk 2]
```diff
@@ -520,8 +520,17 @@ private static bool ContainsFunctionWithWorkerRuntime(IEnumerable<FunctionMetada
             return false;
         }
 
-        public static bool CheckAppOffline(string scriptPath)
+        public static bool CheckAppOffline(IEnvironment environment, string scriptPath)
         {
+            // Linux container environments have an additional way of putting a specific worker instance offline.
+            if (environment.IsLinuxContainerEnvironment())
```
Comment (fabiocav) [inline src/WebJobs.Script/Utility.cs]: nit:combine these checks to avoid nesting

File: src/WebJobs.Script/Utility.cs [hunk 3]
```diff
@@ -520,8 +520,17 @@ private static bool ContainsFunctionWithWorkerRuntime(IEnumerable<FunctionMetada
             return false;
         }
 
-        public static bool CheckAppOffline(string scriptPath)
+        public static bool CheckAppOffline(IEnvironment environment, string scriptPath)
```
Comment (fabiocav) [inline src/WebJobs.Script/Utility.cs]: It would likely make sense to turn this into an environment extension method

File: test/WebJobs.Script.Tests/Rpc/RpcInitializationServiceTests.cs [hunk 4]
```diff
@@ -104,13 +105,44 @@ public async Task RpcInitializationService_Initializes_RpcServerAndChannels_Linu
...
+
             _rpcInitializationService = new RpcInitializationService(_optionsMonitor, mockEnvironment.Object, testRpcServer, _mockLanguageWorkerChannelManager.Object, _logger);
             await _rpcInitializationService.StartAsync(CancellationToken.None);
             _mockLanguageWorkerChannelManager.Verify(m => m.InitializeChannelAsync(LanguageWorkerConstants.PythonLanguageWorkerName), Times.Once);
             _mockLanguageWorkerChannelManager.Verify(m => m.InitializeChannelAsync(LanguageWorkerConstants.NodeLanguageWorkerName), Times.Once);
             _mockLanguageWorkerChannelManager.Verify(m => m.InitializeChannelAsync(LanguageWorkerConstants.JavaLanguageWorkerName), Times.Never);
             Assert.Contains("testserver", testRpcServer.Uri.ToString());
             await testRpcServer.ShutdownAsync();
+
+            FileUtility.Instance = null;
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/Rpc/RpcInitializationServiceTests.cs]: Why these changes?

File: src/WebJobs.Script/Extensions/FileUtility.cs (hunk 1 above)
Comment (balag0) [reply src/WebJobs.Script/Extensions/FileUtility.cs]: Moved to Utility.cs

File: src/WebJobs.Script/Utility.cs (hunk 2 above)
Comment (balag0) [reply src/WebJobs.Script/Utility.cs]: done.

File: src/WebJobs.Script/Utility.cs (hunk 3 above)
Comment (balag0) [reply src/WebJobs.Script/Utility.cs]: done.

File: test/WebJobs.Script.Tests/Rpc/RpcInitializationServiceTests.cs (hunk 4 above)
Comment (balag0) [reply test/WebJobs.Script.Tests/Rpc/RpcInitializationServiceTests.cs]: Wanted to make it explicit that container is not disabled. but works without these changes too. Removed.

File: src/WebJobs.Script.WebHost/Controllers/InstanceController.cs [hunk 5]
```diff
@@ -71,5 +71,23 @@ public IActionResult GetInstanceInfo()
         {
             return Ok(_instanceManager.GetInstanceInfo());
         }
+
+        [HttpPost]
+        [Route("admin/instance/disable")]
+        [Authorize(Policy = PolicyNames.AdminAuthLevel)]
+        public async Task<IActionResult> Disable([FromServices] IScriptHostManager hostManager)
+        {
+            if (_environment.IsLinuxContainerEnvironment())
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: Aren't all the APIs for this controller specific to LinuxContainers?

File: src/WebJobs.Script.WebHost/Controllers/InstanceController.cs [hunk 6]
```diff
@@ -71,5 +71,23 @@ public IActionResult GetInstanceInfo()
...
+            {
+                _logger.LogInformation("Disabling container");
+                // Mark the container disabled. We check for this on host restart
+                await Utility.MarkContainerDisabled();
+                var unused = hostManager.RestartHostAsync();
+                return Ok();
+            }
+
+            _logger.LogWarning("Disable container used in non-linux environment");
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: nit: "Disable can only be used in Linux Container environments"

File: src/WebJobs.Script.WebHost/Controllers/InstanceController.cs [hunk 7]
```diff
@@ -71,5 +71,23 @@ public IActionResult GetInstanceInfo()
...
+        [Route("admin/instance/disable")]
+        [Authorize(Policy = PolicyNames.AdminAuthLevel)]
+        public async Task<IActionResult> Disable([FromServices] IScriptHostManager hostManager)
+        {
+            if (_environment.IsLinuxContainerEnvironment())
+            {
+                _logger.LogInformation("Disabling container");
+                // Mark the container disabled. We check for this on host restart
+                await Utility.MarkContainerDisabled();
+                var unused = hostManager.RestartHostAsync();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: maybe do a Task.Run to run this in the background?

File: src/WebJobs.Script/Rpc/RpcInitializationService.cs [hunk 8]
```diff
@@ -65,8 +65,9 @@ public RpcInitializationService(IOptionsMonitor<ScriptApplicationHostOptions> ap
 
         public async Task StartAsync(CancellationToken cancellationToken)
         {
-            if (Utility.CheckAppOffline(_applicationHostOptions.CurrentValue.ScriptPath))
+            if (Utility.CheckAppOffline(_environment, _applicationHostOptions.CurrentValue.ScriptPath))
             {
+                _logger.LogDebug("AppOffline in RpcInitializationService");
```
Comment (mathewc) [inline src/WebJobs.Script/Rpc/RpcInitializationService.cs]: nit: "App is offline. Skipping Rpc initialization"

File: src/WebJobs.Script/Utility.cs [hunk 9]
```diff
@@ -520,8 +520,28 @@ private static bool ContainsFunctionWithWorkerRuntime(IEnumerable<FunctionMetada
             return false;
         }
 
-        public static bool CheckAppOffline(string scriptPath)
+        public static async Task MarkContainerDisabled()
         {
+            var disableContainerFilePath = Path.Combine(Path.GetTempPath(), ScriptConstants.DisableContainerFileName);
```
Comment (mathewc) [inline src/WebJobs.Script/Utility.cs]: Is the temp directory durable enough for your purposes?

File: src/WebJobs.Script/ScriptConstants.cs [hunk 10]
```diff
@@ -73,6 +73,7 @@ public static class ScriptConstants
         public const string ProxyMetadataFileName = "proxies.json";
         public const string ExtensionsMetadataFileName = "extensions.json";
         public const string AppOfflineFileName = "app_offline.htm";
+        public const string DisableContainerFileName = "disableContainer";
```
Comment (mathewc) [inline src/WebJobs.Script/ScriptConstants.cs]: nit: container_offline maybe?

File: src/WebJobs.Script/Utility.cs [hunk 11]
```diff
@@ -520,8 +520,28 @@ private static bool ContainsFunctionWithWorkerRuntime(IEnumerable<FunctionMetada
...
+            if (!FileUtility.FileExists(disableContainerFilePath))
+            {
+                await FileUtility.WriteAsync(disableContainerFilePath, string.Empty);
```
Comment (mathewc) [inline src/WebJobs.Script/Utility.cs]: Consider writing a value to the file, e.g. "This container instance is offline" or something to that effect

File: src/WebJobs.Script/Utility.cs [hunk 12]
```diff
@@ -520,8 +520,28 @@ private static bool ContainsFunctionWithWorkerRuntime(IEnumerable<FunctionMetada
             return false;
         }
 
-        public static bool CheckAppOffline(string scriptPath)
+        public static async Task MarkContainerDisabled()
```
Comment (mathewc) [inline src/WebJobs.Script/Utility.cs]: Can you write a log statement when this happens?

File: test/WebJobs.Script.Tests.Integration/Management/InstanceControllerTests.cs [hunk 13]
```diff
@@ -80,5 +82,67 @@ public async Task Assign_MSISpecializationFailure_ReturnsError()
...
+
+            FileUtility.Instance = fileSystem.Object;
+
+            scriptHostManager.Setup(s => s.RestartHostAsync(It.IsAny<CancellationToken>()));
+
+            var actionResult = await instanceController.Disable(scriptHostManager.Object);
+
+            FileUtility.Instance = null;
+
+            scriptHostManager.Verify(s => s.RestartHostAsync(It.IsAny<CancellationToken>()), Times.Once);
```
Comment (mathewc) [inline test/WebJobs.Script.Tests.Integration/Management/InstanceControllerTests.cs]: Can you also verify the disable file mock to ensure the file was written?

File: src/WebJobs.Script.WebHost/Controllers/InstanceController.cs (hunk 5 above)
Comment (balag0) [reply src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: Yes. Just making it explicit.

File: src/WebJobs.Script.WebHost/Controllers/InstanceController.cs (hunk 6 above)
Comment (balag0) [reply src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: Fixed.

File: src/WebJobs.Script.WebHost/Controllers/InstanceController.cs (hunk 7 above)
Comment (balag0) [reply src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: following the same pattern in the restart api.

File: src/WebJobs.Script/Rpc/RpcInitializationService.cs (hunk 8 above)
Comment (balag0) [reply src/WebJobs.Script/Rpc/RpcInitializationService.cs]: Fixed.

File: src/WebJobs.Script/Utility.cs (hunk 9 above)
Comment (balag0) [reply src/WebJobs.Script/Utility.cs]: Yes. It is good enough.

Comment (balag0) [reply src/WebJobs.Script/Utility.cs]: Yes.  It works based on my testing.

File: src/WebJobs.Script/ScriptConstants.cs (hunk 10 above)
Comment (balag0) [reply src/WebJobs.Script/ScriptConstants.cs]: Fixed

File: src/WebJobs.Script/Utility.cs (hunk 11 above)
Comment (balag0) [reply src/WebJobs.Script/Utility.cs]: Fixed.

File: src/WebJobs.Script/Utility.cs (hunk 12 above)
Comment (balag0) [reply src/WebJobs.Script/Utility.cs]: Fixed.

File: test/WebJobs.Script.Tests.Integration/Management/InstanceControllerTests.cs (hunk 13 above)
Comment (balag0) [reply test/WebJobs.Script.Tests.Integration/Management/InstanceControllerTests.cs]: Added a assert on the contents of the stream.

File: src/WebJobs.Script.WebHost/Controllers/InstanceController.cs (hunk 7 above)
Comment (balag0) [reply src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: is the concern with this taking too long to respond back?

Comment (mathewc) [reply src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: No, the concern is that this is somewhat of an anti-pattern, since if that Task throws, it will be unhandled and can bring down the process. https://markheath.net/post/async-antipatterns. See everywhere else in the code where we fire/forget async work. We use Task.Run, with exception handling

File: src/WebJobs.Script.WebHost/Controllers/InstanceController.cs (hunk 5 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: Then I'd either remove this specific check because it adds confusion, or make it a controller level filter that applies to all actions.

File: src/WebJobs.Script.WebHost/Controllers/InstanceController.cs (hunk 7 above)
Comment (balag0) [reply src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: Fixed.

File: src/WebJobs.Script/Extensions/FileUtility.cs [hunk 14]
```diff
@@ -282,5 +283,20 @@ private static void DoSafeAction(Action action, bool ignoreErrors)
             {
             }
         }
+
+        public static async Task MarkContainerDisabled(ILogger logger)
+        {
+            logger.LogInformation("Setting container instance offline");
```
Comment (fabiocav) [inline src/WebJobs.Script/Extensions/FileUtility.cs]: Wasn't this being move out of `FileUtility`?

File: src/WebJobs.Script/Extensions/FileUtility.cs [hunk 15]
```diff
@@ -282,5 +283,20 @@ private static void DoSafeAction(Action action, bool ignoreErrors)
...
+            var disableContainerFilePath = Path.Combine(Path.GetTempPath(), ScriptConstants.DisableContainerFileName);
+            if (!Instance.File.Exists(disableContainerFilePath))
+            {
+                await WriteAsync(disableContainerFilePath, "This container instance is offline");
+            }
+        }
+
+        public static bool IsContainerDisabled()
```
Comment (fabiocav) [inline src/WebJobs.Script/Extensions/FileUtility.cs]: This should also be moved from this class

File: src/WebJobs.Script.WebHost/Controllers/InstanceController.cs [hunk 16]
```diff
@@ -71,5 +71,19 @@ public IActionResult GetInstanceInfo()
...
+
+        [HttpPost]
+        [Route("admin/instance/disable")]
+        [Authorize(Policy = PolicyNames.AdminAuthLevel)]
+        public async Task<IActionResult> Disable([FromServices] IScriptHostManager hostManager)
+        {
+            _logger.LogInformation("Disabling container");
+            // Mark the container disabled. We check for this on host restart
+            await Utility.MarkContainerDisabled(_logger);
+#pragma warning disable 4014
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: Consider removing these pragmas and instead just "var tIgnore =  Task.Run(...)"

File: src/WebJobs.Script.WebHost/Controllers/InstanceController.cs [hunk 17]
```diff
@@ -71,5 +71,19 @@ public IActionResult GetInstanceInfo()
         {
             return Ok(_instanceManager.GetInstanceInfo());
         }
+
+        [HttpPost]
+        [Route("admin/instance/disable")]
+        [Authorize(Policy = PolicyNames.AdminAuthLevel)]
+        public async Task<IActionResult> Disable([FromServices] IScriptHostManager hostManager)
+        {
+            _logger.LogInformation("Disabling container");
```
Comment (pragnagopa) [inline src/WebJobs.Script.WebHost/Controllers/InstanceController.cs]: change to logDebug

File: src/WebJobs.Script/OutOfProc/Rpc/RpcInitializationService.cs [hunk 18]
```diff
@@ -65,8 +65,9 @@ public RpcInitializationService(IOptionsMonitor<ScriptApplicationHostOptions> ap
 
         public async Task StartAsync(CancellationToken cancellationToken)
         {
-            if (Utility.CheckAppOffline(_applicationHostOptions.CurrentValue.ScriptPath))
+            if (Utility.CheckAppOffline(_environment, _applicationHostOptions.CurrentValue.ScriptPath))
             {
+                _logger.LogDebug("App is offline. Skipping Rpc initialization");
```
Comment (pragnagopa) [inline src/WebJobs.Script/OutOfProc/Rpc/RpcInitializationService.cs]: nit: `App is offline. RpcInitializationService will not be started`

File: src/WebJobs.Script/Utility.cs [hunk 19]
```diff
@@ -533,8 +533,29 @@ private static bool ContainsFunctionWithWorkerRuntime(IEnumerable<FunctionMetada
             return false;
         }
 
-        public static bool CheckAppOffline(string scriptPath)
+        public static async Task MarkContainerDisabled(ILogger logger)
         {
+            logger.LogInformation("Setting container instance offline");
```
Comment (pragnagopa) [inline src/WebJobs.Script/Utility.cs]: nit: change to logDebug and add two logs, if file exists add log, files exists and if not then, log you have makes sense

File: src/WebJobs.Script/Utility.cs [hunk 20]
```diff
@@ -533,8 +533,29 @@ private static bool ContainsFunctionWithWorkerRuntime(IEnumerable<FunctionMetada
...
+        {
+            // Linux container environments have an additional way of putting a specific worker instance offline.
+            if (environment.IsLinuxConsumptionContainerDisabled())
+            {
+                return true;
+            }
+
             // check if we should be in an offline state
             string offlineFilePath = Path.Combine(scriptPath, ScriptConstants.AppOfflineFileName);
             if (File.Exists(offlineFilePath))
```
Comment (pragnagopa) [inline src/WebJobs.Script/Utility.cs]: change this to use `FileUtility.FileExists` as well

File: test/WebJobs.Script.Tests/Rpc/RpcInitializationServiceTests.cs [hunk 21]
```diff
@@ -104,13 +105,44 @@ public async Task RpcInitializationService_Initializes_RpcServerAndChannels_Linu
             mockEnvironment.Setup(p => p.GetEnvironmentVariable(LanguageWorkerConstants.FunctionWorkerRuntimeSettingName)).Returns(string.Empty);
             mockEnvironment.Setup(p => p.GetEnvironmentVariable(EnvironmentSettingNames.ContainerName)).Returns("LinuxContainer");
 
+            var fileSystem = new Mock<IFileSystem>();
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Rpc/RpcInitializationServiceTests.cs]: move this to constructor

File: test/WebJobs.Script.Tests/Rpc/RpcInitializationServiceTests.cs [hunk 22]
```diff
@@ -104,13 +105,44 @@ public async Task RpcInitializationService_Initializes_RpcServerAndChannels_Linu
...
+                    f.File.Exists(It.Is<string>(path => path.EndsWith(ScriptConstants.DisableContainerFileName))))
+                .Returns(true);
+            FileUtility.Instance = fileSystem.Object;
+
+            _rpcInitializationService = new RpcInitializationService(_optionsMonitor, mockEnvironment.Object, testRpcServer, _mockLanguageWorkerChannelManager.Object, _logger);
+            await _rpcInitializationService.StartAsync(CancellationToken.None);
+            _mockLanguageWorkerChannelManager.Verify(m => m.InitializeChannelAsync(LanguageWorkerConstants.PythonLanguageWorkerName), Times.Never);
+            _mockLanguageWorkerChannelManager.Verify(m => m.InitializeChannelAsync(LanguageWorkerConstants.NodeLanguageWorkerName), Times.Never);
+            _mockLanguageWorkerChannelManager.Verify(m => m.InitializeChannelAsync(LanguageWorkerConstants.JavaLanguageWorkerName), Times.Never);
+            await testRpcServer.ShutdownAsync();
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Rpc/RpcInitializationServiceTests.cs]: add ` Assert.DoesNotContain("testserver", testRpcServer.Uri.ToString());` similar to test: RpcInitializationService_AppOffline  this will verify test server is not started

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs [hunk 23]
```diff
@@ -156,6 +156,16 @@ public static bool IsLinuxConsumption(this IEnvironment environment)
...
         }
 
+        /// <summary>
+        /// Gets a value indicating whether this specific linux consumption container instance is in offline mode.
+        /// </summary>
+        /// <param name="environment">The environment to verify</param>
+        /// <returns><see cref="true"/> if running in a Linux Consumption App Service app and the container is in draining mode; otherwise, false.</returns>
+        public static bool IsLinuxConsumptionContainerDisabled(this IEnvironment environment)
+        {
+            return environment.IsLinuxConsumption() && Utility.IsContainerDisabled();
```
Comment (pragnagopa) [inline src/WebJobs.Script/Environment/EnvironmentExtensions.cs]: Do not add this helper here. Environment extensions should not be using helpers from Utility. Add this check directly in Utility

File: src/WebJobs.Script/Utility.cs [hunk 24]
```diff
@@ -533,8 +533,29 @@ private static bool ContainsFunctionWithWorkerRuntime(IEnumerable<FunctionMetada
...
+                await FileUtility.WriteAsync(disableContainerFilePath, "This container instance is offline");
+            }
+        }
+
+        public static bool IsContainerDisabled()
+        {
+            return FileUtility.FileExists(Path.Combine(Path.GetTempPath(), ScriptConstants.DisableContainerFileName));
+        }
+
+        public static bool CheckAppOffline(IEnvironment environment, string scriptPath)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Utility.cs]: AppOffline and Container disabled are two separate states. Keep them separate

File: src/WebJobs.Script/OutOfProc/Rpc/RpcInitializationService.cs [hunk 25]
```diff
@@ -65,8 +65,9 @@ public RpcInitializationService(IOptionsMonitor<ScriptApplicationHostOptions> ap
 
         public async Task StartAsync(CancellationToken cancellationToken)
         {
-            if (Utility.CheckAppOffline(_applicationHostOptions.CurrentValue.ScriptPath))
+            if (Utility.CheckAppOffline(_environment, _applicationHostOptions.CurrentValue.ScriptPath))
```
Comment (pragnagopa) [inline src/WebJobs.Script/OutOfProc/Rpc/RpcInitializationService.cs]: Add separate checks, AppOffline or ContainerDisabled

File: src/WebJobs.Script/Utility.cs (hunk 19 above)
Comment (balag0) [reply src/WebJobs.Script/Utility.cs]: changed to logdebug. if this is called twice we will have 2 entries for "Setting container instance offline". so we might not need additional logs for that scenario. Also this operation is idempotent and might get called multiple times due to retries on dataservice.

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs (hunk 23 above)
Comment (balag0) [reply src/WebJobs.Script/Environment/EnvironmentExtensions.cs]: Changed it to this way based on Fabio's earlier comment on this PR.

File: test/WebJobs.Script.Tests/Rpc/RpcInitializationServiceTests.cs (hunk 21 above)
Comment (balag0) [reply test/WebJobs.Script.Tests/Rpc/RpcInitializationServiceTests.cs]: Only a few tests need this. so method scoped might be enough.

File: src/WebJobs.Script/Utility.cs (hunk 24 above)
Comment (balag0) [reply src/WebJobs.Script/Utility.cs]: We will have logs to know why the app was put offline and also i think they are just different ways of putting the app offline. so prefer to keep it this way to avoid more code churn.

File: src/WebJobs.Script/OutOfProc/Rpc/RpcInitializationService.cs (hunk 25 above)
Comment (balag0) [reply src/WebJobs.Script/OutOfProc/Rpc/RpcInitializationService.cs]: We will have logs to know why the app was offlined. And also prefer to keep it this way so there is just 1 method to check when deciding to do something for offline apps
